backend/src/server.rs

```diff
@@ -0,0 +1,518 @@
+    for i in 0..data.len() {
+        if skip_list[i] {
+            continue;
+        }
+        let diff = sums[i] - mean;
+        var_total += diff * diff;
+    }
+    let variance = var_total / ((count - 1) as f64);
+
+    let trend = if count >= 10 && count == data.len() {
+        let q1_mean = q1_total / (q1_idx as f64);
+        let q4_mean = q4_total / ((data.len() - q4_idx) as f64);
+        100.0 * ((q4_mean - q1_mean) / first)
+    } else {
+        0.0
+    };
+    let trend_b = 100.0 * ((last - first) / first);
+
+    object! {
+        "first" => first,
+        "last" => last,
+        "min" => min,
+        "max" => max,
+        "mean" => mean,
+        "variance" => variance,
+        "trend" => trend,
+        "trend_b" => trend_b,
+        "n" => count
+    }
+}
+
+/// Pre and post processes the request and response to prepare it for the
+/// handler passed in. Specifically, it parses JSON data from the request if
+/// the request had greater than 0 length, and hands that into the handler.
+/// Post-processing consists of applying access control headers and [potentially]
+/// printing the error message.
+fn get_handler<F>(handler: F, req: &mut Request) -> IronResult<Response>
+    where F: Fn(Option<JsonValue>, &InputData) -> Result<JsonValue> {
+    use std::ops::Deref;
+    use iron::headers::{ContentType, AccessControlAllowOrigin};
+    use iron::mime::{Mime, TopLevel, SubLevel};
+    use iron::modifiers::Header;
+
+    let rwlock = req.get::<State<InputData>>().unwrap();
+    let data = rwlock.read().unwrap();
+
+    let mut buf = String::new();
+    let res = match req.body.read_to_string(&mut buf).unwrap() {
+        0 => handler(None, data.deref()),
+        _ => handler(Some(json::parse(&buf).unwrap()), data.deref())
+    };
+
+    let mut resp = match res {
+        Ok(json) => {
+            let mut resp = Response::with((status::Ok, json.dump()));
+            resp.set_mut(Header(ContentType(Mime(TopLevel::Application, SubLevel::Json, vec![]))));
+            resp
+        },
+        Err(err) => {
+            // TODO: Print to stderr
+            println!("An error occurred: {:?}", err);
+            Response::with((status::InternalServerError, err.to_string()))
+        }
+    };
+    resp.set_mut(Header(AccessControlAllowOrigin::Any));
+    Ok(resp)
+}
+
+pub fn start(data: InputData) {
+    let mut router = Router::new();
+
+    router.get("/summary", |r: &mut Request| get_handler(get_summary, r));
+    router.get("/info", |r: &mut Request| get_handler(get_info, r));
+    router.post("/data", |r: &mut Request| get_handler(get_data, r));
+    router.post("/get_tabular", |r: &mut Request| get_handler(get_tabular, r));
+    router.post("/get", |r: &mut Request| get_handler(get_days, r));
+    router.post("/stats", |r: &mut Request| get_handler(get_stats, r));
+
+    let mut chain = Chain::new(router);
+    chain.link(State::<InputData>::both(data));
+
+    Iron::new(chain).http("0.0.0.0:2346").unwrap();
```

Comment:
I'd probably stick at the top of the crate to make it as easy as possible to change. We'll probably read it from a config file at some point, I guess.